src/dmd/root/filename.d

```diff
@@ -872,6 +879,17 @@ nothrow:
         }
     }
 
+    /// Useful when checking if a file resides in a particular directory
+    /// See_Also:
+    ///     FileName.canonicalName, FileName.exists
+    static inout(char)[] ensureDirSeparatorEnding(inout(char)[] path)
+    {
+        version(Windows)
+            return isDirSeparator(path[$ - 1]) ? path : path ~ '\\';
+        else version(Posix)
+            return isDirSeparator(path[$ - 1]) ? path : path ~ '/';
```

Comment:
FileName functions use mem.xmalloc,xfree and friends for deterministic destruction, so I doubt `~` is allowed here

Reply:
@rainers does this look better?
```d
static const(char)[] ensureDirSeparatorEnding(const(char)[] path)
{
    if (!isDirSeparator(path[$ - 1]))
    {
        char[] result = (cast(char*)mem.xmalloc_noscan(path.length + 1))[0 .. path.length + 1];
        version(Windows)
            result[$ - 1] = '\\';
        else version(Posix)
            result[$ - 1] = '/';
        return result;
    }
    else
        return path;
}
```

Reply:
> @rainers does this look better?

Regarding allocation, yes. But you should also copy `path` to `result`.
BTW: an empty `path` will cause a range error, so better check against that (should yield '/' or './' ?).